src/segment.rs

```diff
@@ -1,39 +1,184 @@
+use crate::print::{Grapheme, UnicodeWidthGraphemes};
 use ansi_term::{ANSIString, Style};
 use std::fmt;
+use unicode_segmentation::UnicodeSegmentation;
 
-/// A segment is a single configurable element in a module. This will usually
-/// contain a data point to provide context for the prompt's user
-/// (e.g. The version that software is running).
+/// Type that holds text with an associated style
```

Comment:
I feel like that comment is not really true even before this change (sometimes a module produces multiple segments that can't be configured independently, sometimes two different elements that share a style might be combined) and definitely isn't correct after this change (because we're actively splitting segments on newlines). 

The thing that seems to be notable about a segment is that it's a chunk of text that has the same styling. It is true that for the most part a segment will be a logical chunk of text, but that's more of a byproduct of sensible styling choices than inherent in the type. 

eg. we could create a format string like `"[r](red)[a](yellow)[i](green)[n](cyan)[b](blue)[o](magenta)[w](white)"` which would be 7 segments but really only one element (a word). This is just uncommon, b/c shell prompts generally use styles for color coding, not just decoration. 

I'm open to expanding on the comment though. How about this?
```
/// A segment is a chunk of text with an associated style.
/// In general, modules should produce one segment for each configurable element in the module. 
/// (eg. The version that software is running might be one element, and the symbol for the module a second segment).
```
(Though TBH, to me it feels like the last two lines really belong as a comment on `Module not on Segment)

Reply:
Totally fair. Go with whatever comment you think is fitting. 👍